File: code/arc_infer.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h>
#include<Eigen/Dense>
#define Pi 3.14159

using namespace Eigen;

double *import_file(int *cnt, char *filename){
	FILE *f;
	f = fopen(filename,"rt");
	*cnt = 0;
	double dp;
	while(fscanf(f, "%lf\n", &dp) != EOF) (*cnt)++;
	rewind(f);
	double *data = (double*) malloc(*cnt*sizeof(double));
	for(int i = 0; i < *cnt; ++i) fscanf(f, "%lf\n", &data[i]);
	fclose(f);
	return data;
}

double ARC_from_sol(double fi, VectorXf sol, int N_fourier){
	double res = sol(1);
	for(int n = 1; n < N_fourier; ++n){
		res += sol(2*n)*cos(n*fi);
		res += sol(2*n+1)*sin(n*fi);
	}
	return res;
}

int main(int argc, char *argv[]){
	
	char *str = argv[1];
	int N_fourier = atoi(str);
	
	// import events, phase and forcing
	int events_cnt, a_events_cnt, amplitude_cnt, phase_cnt, forcing_cnt;
	char filename_e[] = "events/phase_thr_events.txt";
	double *events = import_file(&events_cnt, filename_e);
	char filename_ae[] = "events/amplitude_x0_at_events.txt";
	double *a_events = import_file(&a_events_cnt, filename_ae);
	char filename_a[] = "amplitude/amplitude_x0.txt";
	double *amplitude = import_file(&amplitude_cnt, filename_a);
	char filename_p[] = "phase/phase.txt";
	double *phase = import_file(&phase_cnt, filename_p);
	char filename_f[] = "forcing/forcing.txt";
	double *forcing = import_file(&forcing_cnt, filename_f);
	
	// fill the matrix A
	MatrixXf A(events_cnt-1, 1+1+2*N_fourier+1); 
	for(int i = 0; i < events_cnt-1; ++i){
		// Floquet exponent
		A(i,0) = -amplitude[int(ceil(events[i]))-1]*(ceil(events[i])-events[i]); // first fractional timestep
		for(int t = ceil(events[i]); t < floor(events[i+1]); ++t){
			A(i,0) -= amplitude[t];
		}
		A(i,0) -= amplitude[int(floor(events[i+1]))]*(events[i+1]-floor(events[i+1])); // last fractional timestep
		// Fourier integrals
		A(i,1) = forcing[int(ceil(events[i]))-1]*(ceil(events[i])-events[i]); // first fractional timestep
		for(int t = ceil(events[i]); t < floor(events[i+1]); ++t){
			A(i,1) += forcing[t];
		}
		A(i,1) += forcing[int(floor(events[i+1]))]*(events[i+1]-floor(events[i+1])); // last fractional timestep
		for(int n = 1; n <= N_fourier; ++n){
			// cos
			A(i,2*n) = forcing[int(ceil(events[i]))-1]*cos(n*phase[int(ceil(events[i]))-1])*(ceil(events[i])-events[i]); // first fractional timestep
			for(int t = ceil(events[i]); t < floor(events[i+1]); ++t){
				A(i,2*n) += forcing[t]*cos(n*phase[t]);
			}
			A(i,2*n) += forcing[int(floor(events[i+1]))]*cos(n*phase[int(floor(events[i+1]))])*(events[i+1]-floor(events[i+1])); // last fractional timestep
			// sin
			A(i,2*n+1) = forcing[int(ceil(events[i]))-1]*sin(n*phase[int(ceil(events[i]))-1])*(ceil(events[i])-events[i]); // first fractional timestep
			for(int t = ceil(events[i]); t < floor(events[i+1]); ++t){
				A(i,2*n+1) += forcing[t]*sin(n*phase[t]);
			}
			A(i,2*n+1) += forcing[int(floor(events[i+1]))]*sin(n*phase[int(floor(events[i+1]))])*(events[i+1]-floor(events[i+1])); // last fractional timestep
		}
		// the k*x0 integral for floquet exponent times x0
		A(i,2*N_fourier+2) = ceil(events[i])-events[i]; // first fractional timestep
		for(int t = ceil(events[i]); t < floor(events[i+1]); ++t){
			A(i,2*N_fourier+2) += 1;
		}
		A(i,2*N_fourier+2) += events[i+1]-floor(events[i+1]); // last fractional step	
	}
	// fill the vector b
	VectorXf b(events_cnt-1);
	for(int i = 0; i < events_cnt-1; ++i) b(i) = a_events[i+1]-a_events[i];
	// minimization
	MatrixXf ATAA = A.transpose()*A;
	ATAA = ATAA.inverse();
	ATAA = ATAA*A.transpose();
	VectorXf sol(1+1+2*N_fourier+1);
	sol = ATAA*b;
	
	// write on file
	FILE *f; 
	f = fopen("ARC/sol.txt", "wt");
	for(int i = 0; i < 1+1+2*N_fourier+1; ++i){
		fprintf(f, "%lf\n", sol(i));
	}
	fclose(f);
	
	// estimate x0
	double x0 = sol[2+2*N_fourier]/sol[0]; // k*x0/k
	
	// amplitude recalculation
	double *new_amp = new double[forcing_cnt];
	double *psis = new double[events_cnt];
	for(int t = 0; t < floor(events[0])+1; ++t) new_amp[t] = -1;
	for(int i = 0; i < events_cnt-1; ++i){
		double amp = a_events[i]-x0 + (-sol[0]*(a_events[i]-x0) + ARC_from_sol(0,sol,N_fourier)*forcing[int(ceil(events[i]))-1])*(ceil(events[i])-events[i]); // first fractional timestep
		new_amp[int(ceil(events[i]))] = amp;
		for(int t = ceil(events[i])+1; t < int(floor(events[i+1]))+1; ++t){
			amp = amp + (-sol[0]*amp+ARC_from_sol(phase[t],sol,N_fourier)*forcing[t]);
			new_amp[t] = amp;
		}
		// amplitude at the end
		double psi = amp + (-sol[0]*amp+ARC_from_sol(phase[int(floor(events[i+1]))],sol,N_fourier)*forcing[int(floor(events[i+1]))])*(events[i+1]-floor(events[i+1])); // last fractional timestep
		psis[i] = psi;
		// no rescaling here, it can cause inaccuracies
	}

	// write on file
	f = fopen("amplitude/amplitude.txt", "wt");
	for(int i = 0; i < forcing_cnt; ++i){
		fprintf(f, "%lf\n", new_amp[i]);
	}
	fclose(f);
	f = fopen("amplitude/amplitude_x0.txt", "wt");
	for(int i = 0; i < forcing_cnt; ++i){
		fprintf(f, "%lf\n", new_amp[i]+x0);
	}
	fclose(f);
	
	// error
	double var = 0;
	for(int i = 0; i < events_cnt-1; ++i){
		var += (psis[i]-(a_events[i+1]-x0))*(psis[i]-(a_events[i+1]-x0));
	}
	var /= (a_events_cnt-1);
	double error = sqrt(var);
	// 0 error
	double var0 = 0;
	for(int i = 0; i < events_cnt-1; ++i){
		var0 += (a_events[i]-x0)*(a_events[i]-x0);
	}
	var0 /= (a_events_cnt-1);
	double error0 = sqrt(var0);
	
	// write on file
	f = fopen("error/error_arc.txt", "wt");
	fprintf(f, "%lf\n", error);
	fclose(f);
	f = fopen("error/error0_arc.txt", "wt");
	fprintf(f, "%lf\n", error0);
	fclose(f);
	
	return 0;
	
}

File: code/arc_infer.py
import os
import sys
sys.path.insert(1, os.path.join(sys.path[0], '..')) # perspective from the main directory ../

import parameters
from math import pi, sin, cos, ceil, floor, sqrt
from matplotlib import pyplot
import numpy as np

# run in c
#os.system("g++ -I ./code/ code/arc_infer.c -o code/arc_infer.out")
os.system("./code/arc_infer.out "+str(parameters.prc_N_fourier))

"""
def ARC_from_sol(fi, sol):
	res = float(sol[1])
	for n in range(1, parameters.prc_N_fourier):
		res += float(sol[2*n])*cos(n*fi)
		res += float(sol[2*n+1])*sin(n*fi)
	return res

# import events
events = []
f = open("events/phase_thr_events.txt", "rt")
for line in f.readlines():
	events.append(float(line[0:-1]))
f.close()
events = events[:-1] # one less to avoid out of range
# import amplitude at events
a_at_events = []
f = open("events/amplitude_x0_at_events.txt", "rt")
for line in f.readlines():
	a_at_events.append(float(line[0:-1]))
f.close()
a_at_events = a_at_events[:-1] # one less to avoid out of range
# import amplitude
amplitude = []
f = open("amplitude/amplitude_x0.txt", "rt")
for line in f.readlines():
	amplitude.append(float(line[0:-1]))
f.close()
# import phase
phase = []
f = open("phase/phase.txt", "rt")
for line in f.readlines():
	phase.append(float(line[0:-1]))
f.close()
# import forcing
forcing = []
f = open("forcing/forcing.txt", "rt")
for line in f.readlines():
	forcing.append(float(line[0:-1]))
f.close()

A = [[] for i in range(len(events)-1)]
b = np.matrix([a_at_events[i+1]-a_at_events[i] for i in range(len(events)-1)]).transpose()
for i in range(len(events)-1):
	integral = amplitude[ceil(events[i])-1]*(ceil(events[i])-events[i]) # first fractional timestep
	for t in range(ceil(events[i]),floor(events[i+1])):
		integral += amplitude[t]
	integral += amplitude[floor(events[i+1])]*(events[i+1]-floor(events[i+1])) # last fractional timestep
	A[i].append(-integral) # floquet exponent
	# fourier integrals
	integral = forcing[ceil(events[i])-1]*(ceil(events[i])-events[i]) # first fractional timestep
	for t in range(ceil(events[i]),floor(events[i+1])):
		integral += forcing[t]
	integral += forcing[floor(events[i+1])]*(events[i+1]-floor(events[i+1])) # last fractional timestep
	A[i].append(integral)
	for n in range(1,parameters.prc_N_fourier):
		# cos
		integral = forcing[ceil(events[i])-1]*cos(n*phase[ceil(events[i])-1])*(ceil(events[i])-events[i]) # first fractional timestep
		for t in range(ceil(events[i]),floor(events[i+1])):
			integral += forcing[t]*cos(n*phase[t])
		integral += forcing[floor(events[i+1])]*cos(n*phase[floor(events[i+1])])*(events[i+1]-floor(events[i+1])) # last fractional timestep
		A[i].append(integral)
		# sin
		integral = forcing[ceil(events[i])-1]*sin(n*phase[ceil(events[i])-1])*(ceil(events[i])-events[i]) # first fractional timestep
		for t in range(ceil(events[i]),floor(events[i+1])):
			integral += forcing[t]*sin(n*phase[t])
		integral += forcing[floor(events[i+1])]*sin(n*phase[floor(events[i+1])])*(events[i+1]-floor(events[i+1])) # last fractional timestep
		A[i].append(integral)
	# the k*x0 integral
	integral_x0 = ceil(events[i])-events[i] # first fractional timestep
	for t in range(ceil(events[i]),floor(events[i+1])):
		integral_x0 += 1
	integral_x0 += events[i+1]-floor(events[i+1]) # last fractional timestep
	A[i].append(integral_x0) # floquet exponent time x0

# minimization
A = np.matrix(A)
AT = A.transpose()
ATA = AT*A
ATAinv = np.linalg.inv(ATA)
sol = ATAinv*(AT*b)

# write on file
f = open("ARC/sol.txt", "wt")
for i in range(len(sol)):
	f.write(str(float(sol[i]))+"\n")
f.close()

# plot
#args = [0.1*i for i in range(int(2*pi/0.1))]
#arc = [ARC_from_sol(arg,sol) for arg in args]
#pyplot.plot(args,arc)
#pyplot.show()

# estimate x0
x0 = float(sol[-1])/float(sol[0]) # k*x0/k
#print("x0 = "+str(x0))

# amplitude recalculation
new_amplitude = []
psis = []
for t in range(floor(events[0])+1):
	new_amplitude.append(-1)
for i in range(len(events)-1):
	amp = a_at_events[i]-x0 + (-sol[0]*(a_at_events[i]-x0)+ARC_from_sol(0,sol)*forcing[ceil(events[i])-1])*(ceil(events[i])-events[i]) # first fractional timestep
	new_amplitude.append(amp)
	for t in range(ceil(events[i]),floor(events[i+1])):
		amp = amp + (-sol[0]*amp+ARC_from_sol(phase[t],sol)*forcing[t])
		new_amplitude.append(amp)
	# amplitude at the end
	psi = amp + (-sol[0]*amp+ARC_from_sol(phase[t],sol)*forcing[floor(events[i+1])])*(events[i+1]-floor(events[i+1])) # last fractional timestep
	psis.append(psi)
	# rescale so its a_at_events[i+1]-x0 at the end
	# TURNED THE RESCALLING OFF, I THINK IT MAKES A BETTER FIT
	#for t in range(ceil(events[i]),floor(events[i+1])+1):
		#new_amplitude[t] = new_amplitude[t]*((1-(t-events[i])/(events[i+1]-events[i]))+(t-events[i])/(events[i+1]-events[i])*(a_at_events[i+1]-x0)/psi)

# write on file
f = open("amplitude/amplitude.txt", "wt")
for i in range(len(new_amplitude)):
	f.write(str(float(new_amplitude[i]))+"\n")
f.close()
f = open("amplitude/amplitude_x0.txt", "wt")
for i in range(len(new_amplitude)):
	f.write(str(float(new_amplitude[i])+x0)+"\n")
f.close()

# plot
#pyplot.plot([new_amplitude[i] for i in range(len(new_amplitude))])
#pyplot.plot(forcing,'g')
#for i in range(len(events)):
#	pyplot.plot([events[i],events[i]],[0,a_at_events[i]-x0],'k')
#pyplot.show()

# error
diffs = [(psis[i]-(a_at_events[i+1]-x0))**2 for i in range(len(psis))]
error = sqrt(np.average(diffs))
# 0 error
a_sqs = [(a_at_event-x0)**2 for a_at_event in a_at_events]
error0 = sqrt(np.average(a_sqs))

# write on file
f = open("error/error_arc.txt", "wt")
f.write(str(error)+"\n")
f.close()
f = open("error/error0_arc.txt", "wt")
f.write(str(error0)+"\n")
f.close()
"""
